File: other/code-festival-2015-final-open/D.py
import sys

rl = sys.stdin.readline


class LazySegmentTree:
    def __init__(self, ls: list, segfunc, identity_element=0, lazy_ide=0):
        self.ide_ele = identity_element
        self.lazy_ide_ele = lazy_ide
        self.segfunc = segfunc
        n = len(ls)
        self.N0 = 2 ** (n - 1).bit_length()
        self.data = [self.ide_ele] * (2 * self.N0)
        self.lazy = [self.lazy_ide_ele] * (2 * self.N0)
        
        for i, x in enumerate(ls):
            self.data[i + self.N0 - 1] = x
        for i in range(self.N0 - 2, -1, -1):
            self.data[i] = segfunc(self.data[2 * i + 1], self.data[2 * i + 2])
    
    def gindex(self, left, right):
        L = left + self.N0
        R = right + self.N0
        lm = (L // (L & -L)) >> 1
        rm = (R // (R & -R)) >> 1
        while L < R:
            if R <= rm:
                yield R
            if L <= lm:
                yield L
            L >>= 1
            R >>= 1
        while L:
            yield L
            L >>= 1
    
    def propagates(self, *ids):
        for i in reversed(ids):
            idx = i - 1
            v = self.lazy[idx]
            if v == self.lazy_ide_ele:
                continue
            ################################################################
            self.data[2 * idx + 1] += v
            self.data[2 * idx + 2] += v
            self.lazy[2 * idx + 1] += v
            self.lazy[2 * idx + 2] += v
            ################################################################
            self.lazy[idx] = self.lazy_ide_ele
    
    def update(self, left, right, _x):
        L = self.N0 + left
        R = self.N0 + right
        x = _x
        
        while L < R:
            if R & 1:
                R -= 1
                self.lazy[R - 1] += x
                self.data[R - 1] += x
            if L & 1:
                self.lazy[L - 1] += x
                self.data[L - 1] += x
                L += 1
            L >>= 1
            R >>= 1
        for i in self.gindex(left, right):
            idx = i - 1
            self.data[idx] = self.segfunc(self.data[2 * idx + 1], self.data[2 * idx + 2]) + self.lazy[idx]

    def query(self, left, right):
        self.propagates(*self.gindex(left, right))
        L = left + self.N0
        R = right + self.N0
        res = self.ide_ele
        a, b = [], []
    
        while L < R:
            if L & 1:
                a.append(L - 1)
                L += 1
            if R & 1:
                R -= 1
                b.append(R - 1)
            L >>= 1
            R >>= 1
        for i in a + b[::-1]:
            res = self.segfunc(res, self.data[i])
    
        return res


def solve():
    N = int(rl())
    S, T = [0] * N, [0] * N
    for i in range(N):
        S[i], T[i] = map(lambda x: int(x) - 1, rl().split())
    
    lazy_seg_tree = LazySegmentTree([0] * 10 ** 5, max)
    for i in range(N):
        lazy_seg_tree.update(S[i], T[i], 1)
    
    ans = 10 ** 5
    for i in range(N):
        lazy_seg_tree.update(S[i], T[i], -1)
        ans = min(ans, lazy_seg_tree.query(0, 10 ** 5 + 1))
        lazy_seg_tree.update(S[i], T[i], 1)
    print(ans)


if __name__ == '__main__':
    solve()

File: other/code-festival-2015-final-open/D.nim
import algorithm, deques, heapqueue, math, sets, sequtils, strutils, sugar

proc input(): string =
    return stdin.readLine
proc chmax*[T: SomeNumber](num0: var T, num1: T) =
    num0 = max(num0, num1)
proc chmin*[T: SomeNumber](num0: var T, num1: T) =
    num0 = min(num0, num1)


proc bit_length(n: int): int =
    if n == 0:
      return 0
    let s = toBin(n, 60)
    return 60 - s.find('1')


type
    LazySegmentTree*[T] = ref object
        LV, N0: Natural
        ide_ele, lazy_ide_ele: T
        data, lazy_data: seq[T]
        segfunc: proc (a, b: T): T

proc initLazySegmentTree*[T](size: Natural, ide_ele, lazy_ide_ele: T, f: proc (a, b: T): T, ): LazySegmentTree[T] =
    var
        LV = bit_length(size - 1)
        N0 = 1 shl LV
        data = newSeqWith(2*N0, ide_ele)
        lazy_data = newSeqWith(2*N0, lazy_ide_ele)
    return LazySegmentTree[T](LV: LV, N0: N0, ide_ele: ide_ele, lazy_ide_ele: lazy_ide_ele, data: data, lazy_data: lazy_data, segfunc: f)

iterator gindex*[T](self: var LazySegmentTree[T], left, right: int): int =
    var
        L = (left + self.N0) shr 1
        R = (right + self.N0) shr 1
        lc = if (left and 1) == 1: 0 else: bit_length(L and -L)
        rc = if (right and 1) == 1: 0 else: bit_length(R and -R)
    for i in 0..<self.LV:
        if rc <= i:
            yield R
        if L < R and lc <= i:
            yield L
        L = L shr 1
        R = R shr 1

proc propagates*[T](self: var LazySegmentTree[T], ids: seq[int]) =
    var
        idx: int
        v: T
    for id in reversed(ids):
        idx = id - 1
        v = self.lazy_data[idx]
        if v == self.lazy_ide_ele:
            continue
        self.data[2*idx + 1] += v
        self.data[2*idx + 2] += v
        self.lazy_data[2*idx + 1] += v
        self.lazy_data[2*idx + 2] += v
        self.lazy_data[idx] = self.lazy_ide_ele

proc update*[T](self: var LazySegmentTree[T], left, right: int, x: T) =
    let ids = toSeq(self.gindex(left, right))
    self.propagates(ids)
    var
        L = left + self.N0
        R = right + self.N0
    
    while L < R:
        if (L and 1) == 1:
            self.lazy_data[L - 1] += x
            self.data[L - 1] += x
            inc L
        if (R and 1) == 1:
            dec R
            self.lazy_data[R - 1] += x
            self.data[R - 1] += x
        L = L shr 1
        R = R shr 1
    var idx: int
    for id in ids:
        idx = id - 1
        self.data[idx] = self.segfunc(self.data[2*idx + 1], self.data[2*idx + 2])

proc query*[T](self: var LazySegmentTree[T], left, right: int): T =
    self.propagates(toSeq(self.gindex(left, right)))
    var
        L = left + self.N0
        R = right + self.N0
        res = self.ide_ele
    
    while L < R:
        if (L and 1) == 1:
            res = self.segfunc(res, self.data[L - 1])
            inc L
        if (R and 1) == 1:
            dec R
            res = self.segfunc(res, self.data[R - 1])
        L = L shr 1
        R = R shr 1
    return res


var
    lazy_seg_tree: LazySegmentTree[int]
    S, T: array[10^5, int]


proc solve() =
    var N = input().parseInt
    for i in 0..<N:
        (S[i], T[i]) = input().split.mapIt(it.parseInt - 1)
    
    lazy_seg_tree = initLazySegmentTree(10^5, 0, 0, (a, b) => max(a, b))
    for i in 0..<N:
        lazy_seg_tree.update(S[i], T[i], 1)
    
    var ans = 10^5
    for i in 0..<N:
        lazy_seg_tree.update(S[i], T[i], -1)
        ans.chmin(lazy_seg_tree.query(0, 10^5))
        lazy_seg_tree.update(S[i], T[i], 1)
    echo ans


when is_main_module:
    solve()
